### ISSUE TITLE
jax.numpy argument handling does not work for DeviceArray in lieu of int 

### ISSUE THREAD

**Issue by gnecula** [google/jax#3243]
I just ran into a bug when trying to import into google3 the changes from #3217. One internal test failed because it passes `DeviceArray(-1)` for `axis` in a call to `jax.numpy.expand_dims`. This used to work before #3217. 

The problem is that `lax.numpy.expand_dims` (and many other functions) use code like 
```
if isinstance(axis, int):
    axis = (axis,)
```

This code fails if we pass a DeviceArray(-1) (the failure is [here](https://github.com/google/jax/blob/master/jax/interpreters/xla.py#L943))

I can see how the user may end up with a DeviceArray for that argument, e.g., when they used some on-device computation for it. 

I think we need to handle this systematically. For now, I am going to just make a small fix for `expand_dims` only.

**Comment by hawkinsp** [comment]
Should we just disallow this? I think so. It would require fixing users, but I suspect that's a good change.

**Comment by shoyer** [comment]
I agree, this is worth fixing. The intent of `operator.index()` inside `_canonicalize_axis` was already to allow for safely converting "integer-like" values into `int`, and it should work for scalar values, too.

It looks like NumPy uses equivalent of the following checks to normalize `axis` to a tuple:
```python
# inside squeeze
if not isinstance(axis, tuple):
    axis = (axis,)

# inside expand_dims
if type(axis) not in (tuple, list):
    axis = (axis,)
```

We should probably use one of these in both cases. My inclination would be to pick `if not isinstance(axis, tuple)`. Support for passing in a list is not even part of the documented interface for `np.expand_dims`.

Keep in mind that this is going to break code like `axis=jnp.array([0])`, but given that it's already broken in NumPy I think that's OK. Scalars should definitely be more substitutable for integers than ndarrays are for tuples.

**Comment by gnecula** [comment]
There are actually multiple places in lax.numpy where we check for both `list` and `tuple`. It seems that we are proposing to accept only `tuple`. 

I think it would also make sense to use something like lax.numpy._check_shapelike to give nice error messages in case of misuse.

**Comment by mattjj** [comment]
@gnecula can we close this issue if we just take wherever #3250 wrote `if not isinstance(axis, tuple)` and replace it with `if not isinstance(axis, (tuple, list))`? Or do you have something else in mind?

**Comment by gnecula** [comment]
That would be more consistent, but I was actually thinking of using _check_shapelike (extended to accept a single integer) in all cases where the parameter has to be a sequence of integers. _check_shapelike also looks inside the sequence and gives nicer error messages.

**Comment by jakevdp** [comment]
I think we're in better shape now on this issue, because we generally use `_ensure_index_tuple` which works for both scalar and array inputs, and gives a suitable error when the input is invalid: https://github.com/google/jax/blob/478a95ab74a7b0f964b563f415470dd846316745/jax/_src/api_util.py#L44-L49
